### DISCUSSION TITLE
pse和scroll的halo2怎么做trusted setup的，不是问原理，就是实践中怎么操作的？

### DISCUSSION THREAD

**Discussion by cuiweixie** [z2o-k7e/halo2-step-by-step#45] [Q&A]
pse和scroll的halo2怎么做trusted setup的，不是问原理，就是实践中怎么操作的？

**Comment by Demian101** [comment]
Scroll halo2 后端是 kzg:
- SRS ceremony: https://twitter.com/Scroll_ZKP/status/1614107345956765699

PSE 推动的 Ethereum SRS ceremony(已结束):
- https://ceremony.ethereum.org/
- 结束了，之前你可以不断用鼠标滑动页面生成随机数，然后提交你的 random values

总之，参与的人越多，这个 ceremony 就越可信


之前 ZCash 比较浮夸的 ceremony: 
- destroy the hardware https://petertodd.org/2016/cypherpunk-desert-bus-zcash-trusted-setup-ceremony

**Reply by cuiweixie** [reply]
那如果我基于pse的halo2搞，要怎么做trusted setup?

**Reply by qizhou** [reply]
直接复用已有的trusted setup就好了？

**Reply by Demian101** [reply]
是的，按照 Cecelia 的说法，Prove 的时候自动去下载一个 powers_of_tau 文件，然后 Prove 的时候带着这个 powers_of_tau 去做 commitment

**Comment by Demian101** [comment]
咨询了 taiko、Po 和 even，

Prover Prove 的时候要带着 powers_of_tau 去 Prove

Scroll: https://github.com/scroll-tech/scroll-prover

**Comment by Demian101** [comment]
### TL;DR

Prover Prove 的时候要带着 powers_of_tau 去 Prove

For Halo 2, no trusted setup is required at any point, when using the `halo2_proofs` crate. 
But, If you're using the downstream PSE fork that modified Halo 2 to switch from IPA to KZG, then yes a trusted setup is required.

```rust
// halo2_proofs/src/plonk/prover.rs
let proof = create_proof::<KZGCommitmentScheme<_>, ProverSHPLONK<_>, _, _, _, _>(
        &params,

// rely on :  
// halo2_proofs/src/plonk/prover.rs
let params: ParamsKZG<Bn256> = ParamsKZG::setup(3, OsRng);

// rely on : 
// halo2/halo2_proofs/src/poly/kzg/commitment.rs
impl<E: Engine + Debug> ParamsKZG<E> {
    /// Initializes parameters for the curve, draws toxic secret from given rng.
    /// MUST NOT be used in production.
    pub fn setup<R: RngCore>(k: u32, rng: R) -> Self {
    
        // Calculate g = [G1, [s] G1, [s^2] G1, ..., [s^(n-1)] G1] in parallel.
        let g1 = E::G1Affine::generator();
        let s = <E::Scalar>::random(rng);
        let g2 = <E::G2Affine as PrimeCurveAffine>::generator();
        ///...
        let s_g2 = (g2 * s).into();
```



First of all, about `Setup` : 
### Setup

Let $\mathbb{G}_1$ and $\mathbb{G}_2$ be two elliptic curves with a pairing $e : \mathbb{G}_1 \times \mathbb{G}_2 \rightarrow \mathbb{G}_T$. Let $p$ be the order of $\mathbb{G}_1$ and $\mathbb{G}_2$, and $G$ and $H$ be generators of $\mathbb{G}_1$ and $\mathbb{G}_2$. We will use the shorthand notation

$$[x]_1 = xG \in \mathbb{G}_1 \text{ and } [x]_2 = xH \in \mathbb{G}_2$$

for any $x \in \mathbb{F}_p$.

The trusted setup is a preprocessing step that produces a structured reference string:
$$\mathsf{srs} = ([1]_1, [x]_1, \cdots, [x^{d-1}]_1, [x]_2),$$
where:
- $x \in \mathbb{F}$ is a randomly chosen, **secret** evaluation point; and
- $d$ is the size of the trusted setup, corresponding to the maximum degree polynomial that it can support.

```python
@dataclass
class Setup(object):
    #   ([1]₁, [x]₁, ..., [x^{d-1}]₁)
    # = ( G,    xG,  ...,  x^{d-1}G ), where G is a generator of G_2
    powers_of_x: list[G1Point]
    # [x]₂ = xH, where H is a generator of G_2
    X2: G2Point
```

In this repository, we are using the pairing-friendly [BN254 curve](https://hackmd.io/@jpw/bn254), where:
- `p = 21888242871839275222246405745257275088696311157297823662689037894645226208583`
- $\mathbb{G}_1$ is the curve $y^2 = x^3 + 3$ over $\mathbb{F}_p$;
- $\mathbb{G}_2$ is the twisted curve $y^2 = x^3 + 3/(9+u)$ over $\mathbb{F}_{p^2}$; and
- $\mathbb{G}_T = {\mu}_r \subset \mathbb{F}_{p^{12}}^{\times}$.

We are using an existing setup for $d = 2^{11}$, from this [ceremony](https://github.com/iden3/snarkjs/blob/master/README.md). You can find out more about trusted setup ceremonies [here](https://github.com/weijiekoh/perpetualpowersoftau).

### kzg/commitment.rs - setup()

In `halo2/halo2_proofs/src/poly/kzg/commitment.rs`
 - `let g1 = E::G1Affine::generator();`
 - `let g2 = <E::G2Affine as PrimeCurveAffine>::generator();`

```rust
use halo2curves::pairing::Engine;


/// These are the public parameters for the polynomial commitment scheme.
#[derive(Debug, Clone)]
pub struct ParamsKZG<E: Engine> {
    pub(crate) k: u32,
    pub(crate) n: u64,
    pub(crate) g: Vec<E::G1Affine>,
    pub(crate) g_lagrange: Vec<E::G1Affine>,
    pub(crate) g2: E::G2Affine,
    pub(crate) s_g2: E::G2Affine,
}

impl<E: Engine + Debug> ParamsKZG<E>
where
    E::Scalar: PrimeField,
{
    /// Initializes parameters for the curve, draws toxic secret from given rng.
    /// MUST NOT be used in production.
    pub fn setup<R: RngCore>(k: u32, rng: R) -> Self {
        // Largest root of unity exponent of the Engine is `2^E::Scalar::S`, so we can
        // only support FFTs of polynomials below degree `2^E::Scalar::S`.
        assert!(k <= E::Scalar::S);
        let n: u64 = 1 << k;

        // Calculate g = [G1, [s] G1, [s^2] G1, ..., [s^(n-1)] G1] in parallel.
        let g1 = E::G1Affine::generator();
        let s = <E::Scalar>::random(rng);

        let mut g_projective = vec![E::G1::identity(); n as usize];
        parallelize(&mut g_projective, |g, start| {
            let mut current_g: E::G1 = g1.into();
            current_g *= s.pow_vartime([start as u64]);
            for g in g.iter_mut() {
                *g = current_g;
                current_g *= s;
            }
        });

        let g = {
            let mut g = vec![E::G1Affine::identity(); n as usize];
            parallelize(&mut g, |g, starts| {
                E::G1::batch_normalize(&g_projective[starts..(starts + g.len())], g);
            });
            g
        };

        let mut g_lagrange_projective = vec![E::G1::identity(); n as usize];
        let mut root = E::Scalar::ROOT_OF_UNITY;
        for _ in k..E::Scalar::S {
            root = root.square();
        }
        let n_inv = E::Scalar::from(n)
            .invert()
            .expect("inversion should be ok for n = 1<<k");
        let multiplier = (s.pow_vartime([n]) - E::Scalar::ONE) * n_inv;
        parallelize(&mut g_lagrange_projective, |g, start| {
            for (idx, g) in g.iter_mut().enumerate() {
                let offset = start + idx;
                let root_pow = root.pow_vartime([offset as u64]);
                let scalar = multiplier * root_pow * (s - root_pow).invert().unwrap();
                *g = g1 * scalar;
            }
        });

        let g_lagrange = {
            let mut g_lagrange = vec![E::G1Affine::identity(); n as usize];
            parallelize(&mut g_lagrange, |g_lagrange, start| {
                let end = start + g_lagrange.len();
                E::G1::batch_normalize(&g_lagrange_projective[start..end], g_lagrange);
            });
            drop(g_lagrange_projective);
            g_lagrange
        };

        let g2 = <E::G2Affine as PrimeCurveAffine>::generator();
        let s_g2 = (g2 * s).into();

        Self {
            k,
            n,
            g,
            g_lagrange,
            g2,
            s_g2,
        }
    }
```


#### G1Affine::generator()

In `halo2/halo2_proofs/src/poly/kzg/commitment.rs`
 - `let g1 = E::G1Affine::generator();`
 - `let g2 = <E::G2Affine as PrimeCurveAffine>::generator();`

```rust
use halo2curves::pairing::Engine;

pub struct ParamsKZG<E: Engine> {
    pub(crate) g: Vec<E::G1Affine>,
    pub(crate) g_lagrange: Vec<E::G1Affine>,
    pub(crate) g2: E::G2Affine,
    pub(crate) s_g2: E::G2Affine,
}
```

We go:  https://github.com/privacy-scaling-explorations/halo2curves


#### halo2curves

src/secp256k1/curve.rs
```rust
    let g = Secp256k1::generator();
    let g = Secp256k1::generator();
```

src/lib.rs
```rust
pub use pasta_curves::arithmetic::{Coordinates, CurveAffine, CurveExt};
```
about pasta_curves, go https://github.com/zcash/pasta_curves

#### pasta_curves


### prover.rs - ParamsKZG::setup(3, OsRng)

`halo2_proofs/src/plonk/prover.rs`
 - `params = ParamsKZG::setup(3, OsRng);`
```rust
    let params: ParamsKZG<Bn256> = ParamsKZG::setup(3, OsRng);
    let vk = keygen_vk(&params, &MyCircuit).expect("keygen_vk should not fail");
    let pk = keygen_pk(&params, vk, &MyCircuit).expect("keygen_pk should not fail");
    let mut transcript = Blake2bWrite::<_, _, Challenge255<_>>::init(vec![]);

    // Create proof with wrong number of instances
    let proof = create_proof::<KZGCommitmentScheme<_>, ProverSHPLONK<_>, _, _, _, _>(
        &params,
        &pk,
        &[MyCircuit, MyCircuit],
        &[],
        OsRng,
        &mut transcript,
    );
```
